src/com/github/Kraken3/AFKPGC/LagScanner.java

```diff
@@ -0,0 +1,200 @@
+						break;
+					}
+				}
+			}
+		}
+
+		this.isLagSource = lagSource;
+		this.lagCompute = lagSum;
+		AFKPGC.debug("LagScanner completed centered on ", center, " found lag weight ", lagSum, " after scanning ",
+				chunksTested, " and ", (lagSource) ? "is ": "is not ", " a lag source");
+		AFKPGC.logger.info("Lag test centered on (" + center.getX() + ", " + center.getZ() + ") took " + 
+				(System.currentTimeMillis() - now) + " milliseconds");
+		if (callback != null) {
+			callback.callback(lagSource);
+		}
+	}
+
+	public LagScanner.Result testChunk(Chunk chunk, long now) {
+		String world = chunk.getWorld().getName();
+		long chunkId = (long) chunk.getX() << 32L + (long) chunk.getZ();
+		Map<Long, LagScanner.Result> worldCache = null;
+		if (!LagScanner.cache.containsKey(world)) {
+			worldCache = new HashMap<Long, LagScanner.Result>();
+			cache.put(world, worldCache);
+		} else {
+			worldCache = cache.get(world);
+		}
+
+		LagScanner.Result result = null;
+		if (worldCache.containsKey(chunkId)) {
+			LagScanner.Result cachedResult = worldCache.get(chunkId);
+			if (cachedResult.lastUpdate + cacheTimeout > now) {
+				// hasn't exceeded cache timeout, so use it.
+				result = cachedResult;
+				AFKPGC.debug("Chunk ", chunkId, " found in Cache");
+			} else {
+				// cache expired, remove.
+				worldCache.remove(chunkId);
+				AFKPGC.debug("Chunk ", chunkId, " expired out of Cache");
+			}
+		} // else not cached
+
+		if (result == null) {
+			Map<String, Long> stats = new HashMap<String, Long>();
+			AFKPGC.debug("Chunk ", chunkId, " computing lag contribution");
+			// not in the cache, so let's compute.
+			long totalCost = 0L;
+
+			// Test tiles
+			BlockState[] lagTiles = chunk.getTileEntities();
+			for (BlockState tile : lagTiles) {
+				Material tiletype = tile.getType();
+				long tilecost = (long) LagCostConfig.getInstance().cost(tiletype);
+				totalCost += tilecost;
+				if (stats.containsKey(tiletype.name())) {
+					stats.put(tiletype.name(), tilecost + stats.get(tiletype.name()));
+				} else {
+					stats.put(tiletype.name(), tilecost);
+				}
+				if (totalCost >= lagSourceThreshold) {
+					AFKPGC.debug("Chunk ", chunkId, " passed source threshold just with tiles");
+					break; // if we cross the threshold, don't keep adding.
+				}
+			}
+
+			if (totalCost < lagSourceThreshold) {
+				// Test entities
+				Entity[] lagEntity = chunk.getEntities();
+				for (Entity entity : lagEntity) {
+					EntityType enttype = entity.getType();
+					long entcost = (long) LagCostConfig.getInstance().cost(enttype);
+					totalCost += entcost;
+					if (stats.containsKey(enttype.name())) {
+						stats.put(enttype.name(), entcost + stats.get(enttype.name()));
+					} else {
+						stats.put(enttype.name(), entcost);
+					}
+					if (totalCost >= lagSourceThreshold) {
+						AFKPGC.debug("Chunk ", chunkId, " passed source threshold with tiles and entities");
+						break; // if we cross the threshold, don't keep adding.

```

Comment:
Well, I'm conflicted here. Scanning a full area around a player might take 25-30ms, so not much time, and yes, this shouldn't be run very often. From a "tracking" perspective we shouldn't break out early. From an algorithmic efficiency purpose, we must. Perhaps I could add a config option that determines if it "breaks early" or records the full damage.